token/services/ttx/db.go

```diff
@@ -48,65 +56,132 @@ func (a *DB) Append(tx *Transaction) error {
+	return &TxStatusChangesListener{
+		net:         net,
+		tmsProvider: tmsProvider,
+		tmsID:       tmsID,
+		ttxDB:       ttxDB,
+		tokens:      tokens,
+	}
 }
 
-func (t *TxStatusChangesListener) OnStatusChange(txID string, status int, message string) error {
-	logger.Debugf("tx status changed for tx %s: %s", txID, status)
+func (t *TxStatusChangesListener) OnStatusChange(txID string, status int, statusMessage string) error {
+	defer func() {
+		if e := recover(); e != nil {
+			logger.Errorf("failed on status change [%s]", txID)
+		}
+	}()
+	logger.Debugf("tx status changed for tx [%s]: [%s]", txID, status)
 	var txStatus ttxdb.TxStatus
+	var reference []byte
 	switch network.ValidationCode(status) {
 	case network.Valid:
 		txStatus = ttxdb.Confirmed
+		tokenRequestRaw, err := t.ttxDB.GetTokenRequest(txID)
+		if err != nil {
+			return errors.WithMessagef(err, "failed retriving token request [%s]", txID)
+		}
+		if len(reference) != 0 {
+			tms, err := t.tmsProvider.GetManagementService(token.WithTMSID(t.tmsID))
+			if err != nil {
+				return err
+			}
+			tr, err := tms.NewFullRequestFromBytes(tokenRequestRaw)
+			if err != nil {
+				return err
+			}
+			if err := t.checkTokenRequest(txID, tr, reference); err != nil {
+				// TODO: mark the transaction as invalid
+				return err
+			}
+		}
+		if err := t.tokens.AppendRaw(t.tmsID, txID, tokenRequestRaw); err != nil {
+			// TODO: in this case we have to try again.
+			// at this stage though, we don't fail here because the commit pipeline is processing the tokens still
+			//return errors.WithMessagef(err, "failed to append token request to token db [%s]", txID)
+			logger.Errorf("failed to append token request to token db [%s]: [%s]", txID, err)
+		}
 	case network.Invalid:
 		txStatus = ttxdb.Deleted
 	}
-	if err := t.db.SetStatus(txID, txStatus, message); err != nil {
-		return errors.WithMessagef(err, "failed setting status for request %s", txID)
+
+	if err := t.ttxDB.SetStatus(txID, txStatus, statusMessage); err != nil {
+		return errors.WithMessagef(err, "failed setting status for request [%s]", txID)
 	}
-	logger.Debugf("tx status changed for tx %s: %s done", txID, status)
+	logger.Debugf("tx status changed for tx [%s]: %s done", txID, status)
 	go func() {
-		logger.Debugf("unsubscribe for tx %s...", txID)
+		logger.Debugf("unsubscribe for tx [%s]...", txID)
 		if err := t.net.UnsubscribeTxStatusChanges(txID, t); err != nil {
 			logger.Errorf("failed to unsubscribe auditor tx listener for tx-id [%s]: [%s]", txID, err)
 		}
-		logger.Debugf("unsubscribe for tx %s...done", txID)
+		logger.Debugf("unsubscribe for tx [%s]...done", txID)
 	}()
 	return nil
 }
+
+func (t *TxStatusChangesListener) checkTokenRequest(txID string, request *token.Request, reference []byte) error {
+	trToSign, err := request.MarshalToSign()
+	if err != nil {
+		return errors.Errorf("can't get request hash '%s'", txID)
+	}
+	if base64.StdEncoding.EncodeToString(reference) != hash.Hashable(trToSign).String() {
```

Comment:
Maybe it would make sense to wrap the reference in a struct
```
type TokenReference []byte

func (r *TokenReference) String() {
 return base64.StdEncoding.EncodeToString(*r)
}

func (r *TokenReference) Equals(h Hashable) bool {
 return r.String() == h.String()
}
```